Move the single-expr return transform from Parse to Sema

hamishknight opened the pull request:
Rather than doing the transform in the parser, and then potentially undoing it in Sema, move the entire transform into Sema. This also lets us unify the logic between function decls and closures, and allows ASTGen to benefit from it.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented on lib/AST/Expr.cpp:
```diff
@@ -2034,18 +2037,28 @@ swift::__AbstractClosureExpr_getActorIsolation(AbstractClosureExpr *CE) {
-FORWARD_SOURCE_LOCS_TO(ClosureExpr, Body.getPointer())
+FORWARD_SOURCE_LOCS_TO(ClosureExpr, Body)
 
 Expr *ClosureExpr::getSingleExpressionBody() const {
```
This can now potentially return `true` if a result builder produces a single expression body (which I believe it can do if the original closure is empty), however I don't think that's an issue?

xedin replied:
Right, the builder would have to have `buildBlock() -> ...`. Syntactic element constraint generator has `isSingleExpressionClosure` method that handles result builders as always multi-statement for its purposes and I don't think we care in other places.

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -2664,6 +2633,69 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(
   return false;
 }
 
+/// Insert an implicit return for a single expression body function if needed,
+/// returning \c true if the implicit return was added, \c false otherwise.
```
Need to update this comment

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

slavapestov commented:
> Rather than doing the transform in the parser, and then potentially undoing it in Sema,

We did _what_...? Nice cleanup!

hamishknight commented:
@swift-ci please test source compatibility Debug

hamishknight commented:
@swift-ci please test source compatibility Debug

hamishknight commented:
@swift-ci please test source compatibility Debug

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

ahoppen reviewed (commented):
The code completion changes look good to me (hooray to removing `undoSingleExpressionReturn`).

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/8051

@swift-ci please SourceKit stress test

hamishknight committed:
[test] Update ASTGen tests for single expression returns

These now work because we're doing the transform
in Sema. Closure parameter generation still needs
some extra work unfortunately, so we can't add
tests for that.

hamishknight committed:
[IDE] Remove `undoSingleExpressionReturn`

Now that we do the transform in Sema, this is
redundant.

hamishknight committed:
[CodeComplete] Avoid distinguishing between closure types

This seems to be redundant now that we no longer
attempt to skip multi-statement closure bodies.

hamishknight committed:
[AST] Remove `AbstractFunctionDecl::setSingleExpressionBody`

This is now unused.

hamishknight committed:
Move single-expr return transform from Parse to Sema

Rather than doing the transform in the parser, and
then potentially undoing it in Sema, move the
entire transform into Sema. This also lets us
unify the logic between function decls and
closures, and allows ASTGen to benefit from it.

hamishknight committed:
[AST] Introduce UnreachableExpr

This models the conversion from an uninhabited
value to any type, and allows us to get rid of
a couple of places where we'd attempt to drop
the return statement instead.

hamishknight committed:
[ASTDumper] Remove special case for single expr functions

We don't do the same for closures, and I don't see
any reason why we should obscure the real AST from
the dump.

hamishknight committed:
[AST] Remove the "single expression body" bit

Remove this bit from function decls and closures.
Instead, for closures, infer it from the presence
of a single return or single expression AST node
in the body, which ought to be equivalent, and
automatically takes result builders into
consideration. We can also completely drop this
query from AbstractFunctionDecl, replacing it
instead with a bit on ReturnStmt.

hamishknight committed:
[AST] Remove `AbstractFunctionDecl::getSingleExpressionBody`

This is now unused.

hamishknight committed:
Partially revert "[CodeComplete] Avoid distinguishing between closure types"

This regresses completion performance, bring back
the distinction between the multi-statement
and single-expression fallback cases.

b147a17d755d833d8009ad5caa2f8a856692fdc6

hamishknight commented:
Looks like b1070f20a296503fffd3015b45982832a0605f6a regresses completion performance, reverted it.
